plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs

```diff
@@ -0,0 +1,86 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE TypeFamilies      #-}
+module Language.Plutus.Contract.Contract where
+
+import           Control.Applicative (liftA2)
+import           Control.Monad       ((>=>))
+import           Data.Bifunctor      (first)
+import           Data.List           (foldl')
+
+data Contract t i a =
+    Waiting (i -> Contract t i a)
+    | Done
+    | Emit t (Contract t i a) -- produce a 't' value
+    | Pure a
+    deriving (Functor)
+
+-- The applicative instance parallelises the 'Waiting' operations
+instance Applicative (Contract t i) where
+    pure = Pure
+    cf <*> ca = case cf of
+        Done -> Done
+        Emit t c -> Emit t (c <*> ca)
+        Pure f -> case ca of
+            Done       -> Done
+            Emit t c   -> Emit t (Pure f <*> c)
+            Pure a     -> Pure (f a)
+            Waiting f' -> Waiting $ \i' -> fmap f (f' i')
+        Waiting f ->
+            case ca of
+                Done -> Done
+                Emit t c -> Emit t (Waiting f <*> c)
+                Pure a' -> Waiting $ \i' -> fmap (\f' -> f' a') (f i')
+                Waiting f' ->
+                    Waiting $ \i' -> f i' <*> f' i'
+
+-- The monad instance sequentialises the 'Waiting' operations
+instance Monad (Contract t i) where
+    c >>= f = case c of
+        Waiting f' -> Waiting (f' >=> f)
+        Done       -> Done
+        Pure a     -> f a
+        Emit t c'  -> Emit t (c' >>= f)
+
+instance Semigroup a => Semigroup (Contract t i a) where
+    (<>) = liftA2 (<>)
+
+-- | Apply an input to a contract, collecting as much output data
+--   't' as posible until the contract is blocked on inputs
+applyInput
+    :: Monoid t
+    => i
+    -> Contract t i a
+    -> (t, Contract t i a)
+applyInput ip = \case
+    Waiting f -> drain (f ip)
+    Done -> (mempty, Done)
+    Pure a -> (mempty, Pure a)
+    Emit t c -> first (t <>) (applyInput ip c)
+
+applyInputs
+    :: Monoid t
+    => Contract t i a
+    -> [i]
+    -> ([t], Contract t i a)
+applyInputs c = foldl' go (first return (drain c)) where
+    go (ts, c') i = first (:ts) (applyInput i c')
+
+drain :: Monoid t => Contract t i a -> (t, Contract t i a)
+drain = \case
+    Waiting f -> (mempty, Waiting f)
+    Done -> (mempty, Done)
+    Pure a -> (mempty, Pure a)
+    Emit t c -> first (t <>) (drain c)
+
+outputs :: Monoid t => Contract t i a -> t
+outputs = fst . drain
+
+result :: Monoid t => Contract t i a -> Maybe a
+result = (\case { Pure b -> Just b; _ -> Nothing }) . snd . drain
+
+isDone :: Monoid t => Contract t i a -> Bool
+isDone = (\case { Done -> True; _ -> False }) . snd . drain
+
+await :: t -> (i -> Maybe a) -> Contract t i a
+await a f = Emit a $ Waiting $ maybe (await a f) Pure . f
```

Comment:
I find it odd that this combines emitting with awaiting. Why not add an `emit` combinator?

Also, I'd be inclined to make the type just `await :: Contract t i i`, i.e. it always gives you the input value. This is good enough - we can always turn it into some other type later, e.g.
```
x <- await maybeIToX
```
vs
```
i <- await
let x = maybeIToX i
```

Having written that,  I now see that it calls itself if the maybe-function doesn't return true (I found this quite hard to read, I think it would be better with a pattern match, tbh). I think I'd still be inclined to have something simpler, like `await :: (i -> Bool) -> Contract t i a` or something.

Reply:
The more I think about it the more I think I want
```
await :: Contract t i i
awaitUntil :: (i -> Bool) -> Contract t i i
awaitUntil p = do
    i <- await
    if p i then pure i else awaitUntil p
```

Reply:
I'll add a combinator for each of the constructors of `Contract`. 

But I still think `await` is useful as it is: The type `i` is something like `data Input = LedgerUpdate | SlotChange | Endpoint` so with `i -> Bool` we still have to pattern match on the `i` if the predicate is true and I think it's better to combine that into one. 

That said, I'm not quite happy with the emit/await structure yet. I'll see if I can make it nicer.

Reply:
Note that something like `until` is a self-contained monad combinator, and you can see various people have defined it: https://hoogle.haskell.org/?hoogle=%28a+-%3E+Bool%29+-%3E+m+a+-%3E+m+a&scope=set%3Astackage

There are also versions which have a `m (Maybe a)` which I think corresponds to your current one. I think the version with more combinators is nicer.

Reply:
Yeah I added a `loopM` and used that to define the trigger that watches an address: https://github.com/input-output-hk/plutus/pull/1013/files#diff-0c70358a7c647e9fb38c726c94108085R55